File: database/src/main/java/dreammaker/android/expensetracker/database/Account.java
package dreammaker.android.expensetracker.database;

import com.google.gson.annotations.SerializedName;

import java.math.BigDecimal;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;
import dreammaker.android.expensetracker.database.type.Currency;

import static dreammaker.android.expensetracker.database.ExpensesContract.AboutAccountColumns.BALANCE;
import static dreammaker.android.expensetracker.database.ExpensesContract.AccountsColumns.ACCOUNT_NAME;
import static dreammaker.android.expensetracker.database.ExpensesContract.AccountsColumns._ID;
import static dreammaker.android.expensetracker.database.ExpensesContract.Tables.ACCOUNTS_TABLE;

@Entity(tableName = ACCOUNTS_TABLE)
@Deprecated
public class Account {

    @NonNull
    private String accountName;

    @PrimaryKey(autoGenerate = true)
    private long accountId;

    @NonNull
    private Currency accountBalance;

    @Ignore
    @Deprecated
    public Account(long accountId, String accountName, float balance) {
        this(accountId,accountName,BigDecimal.valueOf(balance));
    }

    @Deprecated
    @Ignore
    public Account(long accountId, String accountName, @NonNull BigDecimal totalBalance) {
        this.accountId = accountId;
        this.accountName = accountName;
        this.accountBalance = Currency.valueOf(totalBalance);
    }

    public Account() {
        this.accountBalance = Currency.ZERO;
    }

    public long getAccountId() { return accountId; }

    public void setAccountId(long accountId) { this.accountId = accountId; }

    @NonNull
    public String getAccountName() { return accountName; }

    public void setAccountName(@NonNull String accountName) { this.accountName = accountName; }

    @Deprecated
    public float getBalance() {
        return getTotalBalance().floatValue();
    }

    @Deprecated
    public void setBalance(float balance) {
        setTotalBalance(BigDecimal.valueOf(balance));
    }

    @NonNull
    public BigDecimal getTotalBalance() {
        return accountBalance.getValue();
    }

    @Deprecated
    public void setTotalBalance(BigDecimal totalBalance) {
        Objects.requireNonNull(totalBalance,"totalBalance == null");
        setAccountBalance(Currency.valueOf(totalBalance));
    }

    @NonNull
    public Currency getAccountBalance() {
        return accountBalance;
    }

    public void setAccountBalance(@NonNull Currency accountBalance) {
        this.accountBalance = accountBalance;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Account)) return false;
        Account account = (Account) o;
        return accountId == account.accountId && accountName.equals(account.accountName) && accountBalance.equals(account.accountBalance);
    }

    @Override
    public int hashCode() {
        return Objects.hash(accountName, accountId, accountBalance);
    }

    @Deprecated
    public boolean equalContents(@Nullable Account o) {
        return equals(o);
    }

    @Override
    public String toString() {
        return "Account{" +
                "accountName='" + accountName + '\'' +
                ", accountId=" + accountId +
                ", balance=" + accountBalance +
                '}';
    }

    @NonNull
    @Override
    public Account clone() {
        Account copy = new Account();
        copy.setAccountId(accountId);
        copy.setAccountName(accountName);
        copy.setAccountBalance(accountBalance);
        return copy;
    }
}

File: database/src/main/java/dreammaker/android/expensetracker/database/account.kt
package dreammaker.android.expensetracker.database

import androidx.lifecycle.LiveData
import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Transaction
import androidx.room.Update
import java.util.Objects

@Entity(tableName = "accounts")
open class Account (
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    var accountId: Long,
    @ColumnInfo(name = "account_name")
    var accountName: String,
    open var balance: Float
): Cloneable {

    override fun equals(other: Any?): Boolean {
        if (other == this) return true
        if (other is Account) {
            return other.accountId == accountId && other.accountName == accountName && other.balance == balance;
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(accountId,accountName,balance)

    @Deprecated("", ReplaceWith("equals(o)"))
    fun equalContents(o: Account?): Boolean = equals(o)

    override fun toString(): String {
        return "Account{" +
                "accountName='" + accountName + '\'' +
                ", accountId=" + accountId +
                ", balance=" + balance +
                '}'
    }

    public override fun clone(): Account = Account(accountId, accountName, balance)
}

data class AccountModel(
    @ColumnInfo(name = "_id")
    val id: Long?,
    @ColumnInfo(name = "account_name")
    val name: String?,
    val balance: Float?
) {
    fun toAccount(): Account {
        return Account(
            id ?: 0,
            name!!,
            balance!!
        )
    }
}

@Dao
interface AccountDao {

    @Insert
    fun insertAccount(account: Account): Long

    @Update
    fun updateAccount(account: Account)

    @Query("SELECT `_id`, `account_name`, `balance` FROM `accounts` ORDER BY `account_name` ASC")
    fun getAllAccounts(): LiveData<List<AccountModel>>

    @Query("SELECT * FROM `accounts` WHERE `_id` = :id")
    fun findAccountById(id: Long): LiveData<AccountModel?>

    @Query("SELECT SUM(`balance`) AS `total_balance` FROM `accounts`")
    fun getTotalBalance(): LiveData<Double?>

    @Delete
    fun deleteAccount(account: Account)

    @Query("DELETE FROM `accounts` WHERE `_id` IN(:ids)")
    fun deleteMultipleAccount(ids: List<Long>)

    @Query("SELECT `_id`, `account_name`, `balance` FROM `accounts` WHERE `_id` IN " +
            "(SELECT `primaryAccountId` FROM `histories` WHERE `primaryAccountId` IS NOT NULL GROUP BY primaryAccountId ORDER BY Max(`date`) DESC LIMIT 3)")
    fun getLatestUsedThreeAccounts(): LiveData<List<AccountModel>>

    @Insert
    @Transaction
    fun insertAccounts(accounts: List<Account>)

    @Query("SELECT * FROM `accounts`")
    fun getAccounts(): List<Account>
}
